src/ray/gcs/gcs_server/gcs_actor_scheduler.h

```diff
@@ -0,0 +1,235 @@
+    ActorID GetAssignedActorID() const { return assigned_actor_id_; }
+
+    /// Get the leased resources.
+    const std::vector<rpc::ResourceMapEntry> &GetLeasedResources() const {
+      return resources_;
+    }
+
+   protected:
+    /// The address of the remote leased worker.
+    rpc::Address address_;
+    /// The resources leased from remote node.
+    std::vector<rpc::ResourceMapEntry> resources_;
+    /// Id of the actor assigned to this worker.
+    ActorID assigned_actor_id_;
+  };
+
+ public:
+  /// Create a GcsActorScheduler
+  ///
+  /// \param io_context The main event loop.
+  /// \param actor_info_accessor Used to flush actor info to storage.
+  /// \param gcs_node_manager The node manager which is used when scheduling.
+  /// \param schedule_failure_handler Invoked when there are no available nodes to
+  /// schedule actors.
+  /// \param schedule_success_handler Invoked when actors are created on the worker
+  /// successfully.
+  /// \param lease_client_factory Factory to create remote lease client, default factor
+  /// will be used if not set.
+  /// \param client_factory Factory to create remote core worker client, default factor
+  /// will be used if not set.
+  explicit GcsActorScheduler(
+      boost::asio::io_context &io_context, gcs::ActorInfoAccessor &actor_info_accessor,
+      const GcsNodeManager &gcs_node_manager,
+      std::function<void(std::shared_ptr<GcsActor>)> schedule_failure_handler,
+      std::function<void(std::shared_ptr<GcsActor>)> schedule_success_handler,
+      LeaseClientFactoryFn lease_client_factory = nullptr,
+      rpc::ClientFactoryFn client_factory = nullptr);
+  virtual ~GcsActorScheduler() = default;
+
+  /// Schedule the specified actor.
+  /// If there is no available nodes then the `schedule_failed_handler_` will be
+  /// triggered, otherwise the actor will be scheduled until succeed or canceled.
+  ///
+  /// \param actor to be scheduled.
+  void Schedule(std::shared_ptr<GcsActor> actor);
+
+  /// Cancel all actors that are being scheduled to the specified node.
+  ///
+  /// \param node_id ID of the node where the worker is located.
+  /// \return ID list of actors associated with the specified node id.
+  std::vector<ActorID> CancelOnNode(const ClientID &node_id);
+
+  /// Cancel the actor that is being scheduled to the specified worker.
+  ///
+  /// \param node_id ID of the node where the worker is located.
+  /// \param worker_id ID of the worker that the actor is creating on.
+  /// \return ID of actor associated with the specified node id and worker id.
+  ActorID CancelOnWorker(const ClientID &node_id, const WorkerID &worker_id);
+
+  /// Get leased workers in phase of creating.
+  /// This method is used in unit test.
+  const absl::flat_hash_map<
+      ClientID, absl::flat_hash_map<WorkerID, std::shared_ptr<GcsLeasedWorker>>>
+      &GetWorkersInPhaseOfCreating() const {
+    return node_to_workers_when_creating_;
+  }
+
+ protected:
```

Comment:
Is there a reason these methods/fields protected instead of private? If there isn't, can you make them private?

Reply:
> Is there a reason these methods/fields protected instead of private? If there isn't, can you make them private?

I want to mock this class in unit test, so these member functions are defined as protected so that they can be accessed directly in unit test.